web/api/tests/valid_urls.c

```diff
@@ -0,0 +1,771 @@
+                                            .rwlock = AVL_LOCK_INITIALIZER } };
+
+/* Note: this is not a CMocka group_test_setup/teardown pair. This is performed per-test.
+*/
+static struct web_client *setup_fresh_web_client()
+{
+    struct web_client *w = (struct web_client *)malloc(sizeof(struct web_client));
+    memset(w, 0, sizeof(struct web_client));
+    w->response.data = buffer_create(NETDATA_WEB_RESPONSE_INITIAL_SIZE);
+    w->response.header = buffer_create(NETDATA_WEB_RESPONSE_HEADER_SIZE);
+    w->response.header_output = buffer_create(NETDATA_WEB_RESPONSE_HEADER_SIZE);
+    strcpy(w->origin, "*"); // Simulate web_client_create_on_fd()
+    w->cookie1[0] = 0;      // Simulate web_client_create_on_fd()
+    w->cookie2[0] = 0;      // Simulate web_client_create_on_fd()
+    w->acl = 0x1f;          // Everything on
+    return w;
+}
+
+static void destroy_web_client(struct web_client *w)
+{
+    buffer_free(w->response.data);
+    buffer_free(w->response.header);
+    buffer_free(w->response.header_output);
+    free(w);
+}
+
+//////////////////////////// Test cases ///////////////////////////////////////////////////////////////////////////////
+
+static void only_root(void **state)
+{
+    (void)state;
+
+    if (localhost != NULL)
+        free(localhost);
+    localhost = malloc(sizeof(RRDHOST));
+
+    struct web_client *w = setup_fresh_web_client();
+    buffer_strcat(w->response.data, "GET / HTTP/1.1\r\n\r\n");
+
+    char debug[4096];
+    repr(debug, sizeof(debug), w->response.data->buffer, w->response.data->len);
+    printf("-> \"%s\"\n", debug);
+
+    //char expected_url_repr[4096];
+    //repr(expected_url_repr, sizeof(expected_url_repr), def->url_out_repr, strlen(def->url_out_repr));
+
+    expect_string(__wrap_mysendfile, filename, "/");
+
+    web_client_process_request(w);
+
+    //assert_string_equal(w->decoded_query_string, def->query_out);
+    destroy_web_client(w);
+    free(localhost);
+    localhost = NULL;
+}
+
+static void two_slashes(void **state)
+{
+    (void)state;
+
+    if (localhost != NULL)
+        free(localhost);
+    localhost = malloc(sizeof(RRDHOST));
+
+    struct web_client *w = setup_fresh_web_client();
+    buffer_strcat(w->response.data, "GET // HTTP/1.1\r\n\r\n");
+
+    char debug[4096];
+    repr(debug, sizeof(debug), w->response.data->buffer, w->response.data->len);
+    printf("-> \"%s\"\n", debug);
+
+    //char expected_url_repr[4096];
+    //repr(expected_url_repr, sizeof(expected_url_repr), def->url_out_repr, strlen(def->url_out_repr));
+
+    expect_string(__wrap_mysendfile, filename, "//");
+
+    web_client_process_request(w);
+
+    //assert_string_equal(w->decoded_query_string, def->query_out);
+    destroy_web_client(w);
+    free(localhost);
+    localhost = NULL;
+}
+
+static void absolute_url(void **state)
+{
+    (void)state;
+
+    if (localhost != NULL)
+        free(localhost);
+    localhost = malloc(sizeof(RRDHOST));
```

Comment:
At netdata we create localhost only when it starts, here you are creating for the second time inside this file, cannot this give the wrong vision about what netdata does?

Reply:
Another difference, we allocate it with callocz instead mallocz.

Reply:
I think unit testing code operates under different constraints to the main application code. This code is not intended to be executed at "run-time" for the application, only at "build-time". The reasoning behind flushing out the `localhost` allocation is to guarantee that we don't have state leakage between tests. We could think of each individual test as an application lifetime of its own, so we do things multiple times within a test-driver that we would only do once within the application itself.

The lack of `_z` calls is to prevent code in the system being tested being called as part of the testing environment itself. In this case I think the l`ocalhost` is unused but we need a dummy parameter for some of the procedures being called.

Reply:
Thank you for the explanation!